### ISSUE TITLE
Separate internal data model from specific serialization formats

### ISSUE THREAD

**Issue by maxnoe** [gammapy/gammapy#3767]
Currently, gammapy's internal data structures are very deeply intertwined with the first version(s) of the GADF specification. 

Mostly, data is read as is from FITS into astropy tables and then throughout the code, the data originally stored in the FITS headers is accessed via `table.meta`.

This has several problems:

* No easy support of multiple formats, since the internal data structure is tied to one specific format
* Data is not in the optimal in-memory representation, some examples:
   * observation begin / end (TSTART / TSTOP) should be `astropy.time.Time` instances, not the internal datetime format of the DL3 files
   * quantities in general are pure numbers, not `astropy.units.Quantity`
* Data is not validated on input. Errors can happen deep into the code for something that could have been caught on input file reading.
* Writing data out is harder, because no clear assumptions can be made about the data in `.meta`: it's a mixture of what existed in the input file (if any) and what is set by the user and what is created by gammapy.

We should define the internal data model, via the corresponding data classes (EventList, IRFs, etc. ) and have a clear IO boundary between internal and external data representations.

**Comment by maxnoe** [comment]
Starting the evaluation of what is required by gammapy and how that would be best represented in memory and where that information is currently read from 

### Observation

The current observation relies mostly on the `obs_info` metadata dict, that is taken either from the header of the `EVENTS` table when directly reading a single observation or it is taken from the `OBS-INDEX` table.

As discussed in the data model meeting, this is problematic for two reasons:
* This information is mandatory in the `EVENTS` header but optional in the `OBS-INDEX` table. Right now, in the standard analysis use case of gammapy, going through datastore, the information is only accessed from the optional location.
* There might be inconsistencies between OBS-INDEX and EVENTS header, then the event header should be the one source of truth.


Looking at the places in gammapy that use obs-info, I identified these properties are used in the analysis:

### Observatory location

Read from GEOLON / GEOLAT / (GEOALT or ALTITUDE).

Note: this is incompatible with the current FITS standard and should be expanded to also support the current standard using the `OBSGEO-{X,Y,Z}` keywords:

> The location, if specified, shall be represented as a geocentric Cartesian triple with respect to a standard ellipsoidal geoid at the time of the observation

Or alternatively:

> A set of geodetic coordinates is recognized by the following keywords OBSGEO-{B,L,H}

The in memory representation should be a `astropy.coordinates.EarthLocation`.

It seems the location is currently not used during the analysis. Which is probably due to the fact that only observations tracking a fixed sky position are supported.

To support other types of observations, e.g. drift mode like HAWC, the observatory location would be needed to correctly stack events in skyfixed coordinates.

### Pointing-Direction

Pointing direction is currently read from the `RA_PNT` and `DEC_PNT` obs_info keywords for equatorial coordinates and `ALT_PNT` / `AZ_PNT` for horizontal coordinates.

The code is also duplicated. A more naive implementation sits directly on the `Observation` object, a more correct implementation at `FixedPointingInfo`.

Both sets of coordinates are assumed to not vary within the run.
This is not a valid assumption in almost all cases, since at least one of the two representations will vary with time.

This also makes no use of the `POINTING` HDU that is present in the GADF.

The properties construct astropy `SkyCoord` from  these meta information, but that has two issues currently:
* The reference frame is not taken into account, always assuming ICRS for equatorial, ignoring `RADECSYS`.
* The horizontal coordinate is constructed using an incomplete frame. Time and observatory location are missing and the coordinate can thus not be used for transformations, only for looking at the `AltAz` values.

I think the correct way to deal with this would be to:

* Also look for the `POINTING` table
* Add methods `get_pointing_equatorial(t: Time)` and `get_pointing_horizontal(t: Time)` to honor the fact of changing pointings returning `SkyCoord` with completely initalized frame attributes.
* POINTING table should take precedence over header keywords in the EVENTS, which are not well-defined and probably contain averages for the coordinates that change during the observation. 


### Dead time fraction / Livetime / Duration

Taken from the `LIVETIME` / `TSTART` / `TSTOP` / `DEADC` / `ONTIME`.

The observation needs to store `ontime` or `livetime` plus `deadc`.

Start and stop time should be `astropy.time.Time` and ontime or livetime either a `TimeDelta` or a `Quantity` in seconds,
the latter is probably preferred.

### Unused properties

There are a number of properties that are currently unused in the analysis code:

* muoneff → obs_info.get("MUONEFF", 1)
* target_radec
* pointing_radec, pointing_altaz, pointing_zen, see above discussion on pointing information duplication.

Target ra dec is important metadata, but as serious limitations. What if multiple sources are in the FoV? What if the Target is known to be in the FoV but at an uncertain position? It is currently not used in gammapy, since users define their regions of interest themselves independent of this.


### API Proposal

I think an observation should have these attributes / properties / methods


class Observation:
     location: EarthLocation
     t_start: Time
     t_stop: Time
     livetime: Quantity(unit=s)
     get_pointing_equatorial(t: Time)
     get_pointing_altaz(t: Time)
     mode: Enum(...)
     meta: dict()  # for the actual, user defined metadata.


Anything important missing?

Do we want/need to support satellite data here?
https://docs.astropy.org/en/stable/coordinates/satellites.html

According to the fits standard, the solution is then to provide an `OBSORBIT` containing either a url or a filename for orbital ephemeris instead.


A question that will be important to answer if we are to support multiple formats, not just multiple versions of FITS based formats, is if we want to define common metadata in a more structured way. I am thinking about:

* Observatory name
* Instrument name
* Organisation
* Software
....

**Comment by maxnoe** [comment]
### EventList

EventList currently directly contains the table as read using `Table.read` from the input file. This should be separated.

We could still use a table or define array-like attributes directly on the EventList like this:


```
class EventList:
    reconstructed_position: SkyCoord
    time: Time
    reconstructed_energy: Quantity[TeV]

    # optional columns for more, e.g. true quantities?
    # maybe just keep the original table?
```

the current GADF standard requires all kinds of observation metadata to be defined on the `EVENTS` table,
but in the current schema, these would belong to `Observation` in the in-memory representation of gammapy, which I think makes sense. So either the `Event` could just have a backref to the observation, or the write method could take that as an additional argument like that:

```
events.write('events.dl3.fits.gz', format='gadf-dl3', version='0.2', observation=observation)
```

**Comment by maxnoe** [comment]
### IRFs

We need 

* is_pointlike for everything but the PSF (Background is arguable, you could provide point-like background estimates, but this is uncommon I think)
* fov_alignment for IRFs that depend on DETX/DETY (Currently only 3D Background)

The rest should be covered by the properties we already have, e.g. interpolation should be read into the axes definition.

**Comment by adonath** [comment]
> It seems the location is currently not used during the analysis. Which is probably due to the fact that only observations tracking a fixed sky position are supported.

It is used here https://github.com/gammapy/gammapy/blob/master/gammapy/makers/map.py#L163 The information currently translates into a `FixedPointingInfo` https://github.com/gammapy/gammapy/blob/master/gammapy/data/pointing.py#L14

**Comment by maxnoe** [comment]
@adonath Indeed, but that goes again through fixed_pointing_info, not using the location property on the observation.
